### ISSUE TITLE
Regression: no-unused-variable misbehaving for generic types

### ISSUE THREAD

**Issue by joscha** [palantir/tslint#2621]
### Bug Report

- __TSLint version__: 5.1.0
- __TypeScript version__: 2.2.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
type Props = {
    foo: string,
};

export class X extends React.Component<Props, void> {}
```

with `tslint.json` configuration:

```json
{
    "extends": "tslint:latest",
    "defaultSeverity": "error",
    "rules": {
        "interface-over-type-literal": false,
        "no-unused-variable": true
    }
}
```

seet gist here for a full example: https://gist.github.com/joscha/6633bae73fb4b143cfb685b2754259c9

run `yarn install && yarn test` to see the regression.

#### Actual behavior

```
yarn test v0.23.2
$ tslint  --type-check --project tsconfig.json index.ts

ERROR: index.ts[3, 6]: 'Props' is declared but never used.
error Command failed with exit code 2.
```

#### Expected behavior

`Props` is used in the type definition of the React component. This worked with TSLint 4.x.

**Comment by adidahiya** [comment]
I think this is the same as #2470. There was a breaking change in v5.0 where we switched to using `tsc`'s implementation of "no unused variable".
